pkg/runtime/runtime.go

```diff
@@ -2620,10 +2623,10 @@ func (a *DaprRuntime) Shutdown(duration time.Duration) {
 		}
```

Comment:
Shutdown for trace is called with a background context and I have tested it to close neatly even when no output components were setup.

Reply:
The problem isn't the background context (whcih is just a "cancellation token") but that there's not much time for this to work in background now, since there's no "time.After" after this anymore. So it's like a "race against time".

Maybe move it out of a background goroutine?

Reply:
It's at least a possibility, yes. If you keep it in background, I'd add a wait group with a delay. Or if you want to move it to foreground, maybe add a timeout to the context.

Reply:
The current implementation in runtime does not capture tracerProvider post registry. All validations related to trace shutdown require this handler.Fixed the trace interface method, trace initiation, close and added tests to validate closure of trace.

https://github.com/dapr/dapr/blob/4d350a72191b113ea43741f3a9f639fc2f4b20d7/pkg/runtime/runtime.go#L409-L411


 ForceFlush() helps push out all the completed spans before closure of the trace. Earlier this shutdown was dead code since a.tracerProvider was always nil. The shutdown method would always fail with context.Cancelled error. Triggering ForceFlush and Shutdown with a.ctx helps give more time for the trace to gracefully exit. Shutdown method is deferred and not called until exit to allow forceFlush and shutdown to complete.

https://github.com/dapr/dapr/blob/4d350a72191b113ea43741f3a9f639fc2f4b20d7/pkg/runtime/runtime.go#L2615-L2630

@yaron2 @ItalyPaleAle @mukundansundar